[refactor] Make TransactionWithinChainLimit more flexible

kallewoof opened the pull request:
Currently, `TransactionWithinChainLimit` is restricted to single-output use, and needs to be called every time for different limits. If it is replaced with a chain limit value calculator, that can be called once and reused, and is generally more flexible (see e.g. #12257).

Update: this PR now corrects usage of max ancestors / max descendants, including calculating the correct max descendant value, as advertised for the two limits.

~~This change also makes `nMaxAncestors` signed, as the replacement method will return `-1` for "not in the mempool", which is different from "0", which means "no ancestors/descendants in mempool".~~

~~This is a subset of #12257.~~

kallewoof committed:
Remove deprecated TransactionWithinChainLimit

promag commented on src/txmempool.cpp:
```diff
@@ -1055,11 +1055,10 @@ void CTxMemPool::TrimToSize(size_t sizelimit, std::vector<COutPoint>* pvNoSpends
 }
 
-bool CTxMemPool::TransactionWithinChainLimit(const uint256& txid, size_t chainLimit) const {
+int64_t CTxMemPool::chain_limit_value(const uint256& txid) const {
```
Snake case is for variables.

kallewoof replied:
It seems like we are moving to snake case, e.g. for new code like

https://github.com/bitcoin/bitcoin/blob/a34ac6ae0788b50e12dd2c8ac59dbda2a03a3c2e/src/cuckoocache.h#L92

Maybe I'm mistaken on that one. Will check recent merges.

promag replied:
From the developer notes:
  - Variable and namespace names are all lowercase, and may use `_` to
    separate words (snake_case).
  - Class names, function names and method names are UpperCamelCase
    (PascalCase).

kallewoof replied:
Wow, okay. I totally missed that, thanks!

kallewoof replied:
Fixed.

promag commented on src/txmempool.h:
```diff
@@ -619,8 +619,10 @@ class CTxMemPool
-    /** Returns false if the transaction is in the mempool and not within the chain limit specified. */
-    bool TransactionWithinChainLimit(const uint256& txid, size_t chainLimit) const;
+    /** Get the maximum of the chain limit value in the mempool of the given transaction.
+      * If the transaction is not in the mempool, -1 is returned.
```
if not in the mempool could return 0 since `GetCountWithAncestors` and `GetCountWithDescendants` are always `> 0` (self is counted). In that case you can revert s/uint64_t/int64_t.

NicolasDorier replied:
I looked in the code and it seems to me they can be both 0 and that self is not counted.

promag replied:
@NicolasDorier have you checked these?
https://github.com/bitcoin/bitcoin/blob/a34ac6ae0788b50e12dd2c8ac59dbda2a03a3c2e/src/txmempool.cpp#L30
https://github.com/bitcoin/bitcoin/blob/a34ac6ae0788b50e12dd2c8ac59dbda2a03a3c2e/src/txmempool.cpp#L36
https://github.com/bitcoin/bitcoin/blob/a34ac6ae0788b50e12dd2c8ac59dbda2a03a3c2e/src/txmempool.cpp#L317
https://github.com/bitcoin/bitcoin/blob/a34ac6ae0788b50e12dd2c8ac59dbda2a03a3c2e/src/txmempool.cpp#L326

NicolasDorier replied:
woops, completely missed that indeed. I failed to notice how the clients were forcing this invariant though.

NicolasDorier replied:
Got confused when I saw [that](https://github.com/bitcoin/bitcoin/blob/9501dc27b336fc20adb0c367e0904b57bd507f51/src/txmempool.cpp#L857) which seemed to me it could be 0.

kallewoof replied:
I made the change for a reason but that reason may have been flawed. I'll try it with uint64_t and 0 again. Thanks!

kallewoof replied:
The thing is, the check in `SelectCoinsMinConf` checks if the chain limit value `>= nMaxAncestors`, which is `0` for `no chained ancestors` in the code. With the old method, this would be true, but with a 0 returned, it would be false.

I could change the `0` to a `1` to fix it, which may be a better solution here, and is probably what I will end up doing.

kallewoof replied:
If I do that, the semantics becomes misleading. It says 'max *ancestors*' so 1 makes it sound like it allows 1 ancestor or descendant, when in reality it only counts itself. I'm gonna keep the change to signed, for now.

NicolasDorier replied:
Indeed, now:

with @promag suggestion,  if `nMaxAncestors==0` and the tx is not in mempool then `mempool.ChainLimitValue(pcoin->GetHash()) >= nMaxAncestors` evaluate to `true`.

While before `!mempool.TransactionWithinChainLimit(pcoin->GetHash(), nMaxAncestors)` evaluate to `false`.

promag reviewed (commented):
> which is different from "0", which means "no ancestors/descendants in mempool".

I believe "1" means "no ancestors/descendants in mempool".

kallewoof committed:
Add GetTransactionAncestry to CTxMemPool for general purpose chain limit checking

kallewoof committed:
Switch to GetTransactionAncestry() in OutputEligibleForSpending

NicolasDorier commented:
utACK b0928df

promag commented:
utACK 8154223.

jimpo commented:
I find this whole thing unnecessarily confusing. Not your change in particular, more how the code works today.

There are two separate config limits, `-limitancestorcount` and `-limitdescendantcount`, which are treated as the same and merged into a single limit `nMaxChainLength`. Then that value is compared against both the ancestor count and descendant count of the transaction containing the candidate output. So first, it seems like there should be a separate `max_descendants` field on EligibilityFilter.

But further, the descendants check isn't even right: the documentation says "Do not accept transactions if any ancestor would have <n> or more in-mempool descendants". So the descendants check really needs to check the descendants count on all ancestors, not just the direct parent.

Regardless of all that, even just looking at the ancestors limit, which is easier to check, I think it should not be a strict equality check, whereas it is currently in the code. If trying to spend an output from a transaction with 3 ancestors, the spending transaction would have 4 ancestors. Say the ancestor limit is 3. The code currently would call that spend ineligible because GetCountWithAncestors returns 3, and there is a strict < check, whereas the spend should be allowed because the ancestor count is not above the limit. So that is to say `TransactionLimitValue` can return 0 on the case where the tx is not in the mempool.

kallewoof commented:
@jimpo Yeah it's not entirely straightforward that's for sure, and I think further consideration should be given. I don't think this PR makes the situation worse, though.

jimpo commented:
Well, I think it does because it formalizes the concept of a `chain_limit_value`, which is a broken concept. If such a value were useful, it should really be two things, the count of ancestors and the max count of descendants of any ancestor. But furthermore, to calculate them (especially the latter), it's helpful to have an explicit limit on the depth of recursive searching, which is why they are passed into the mempool function rather than returned from it. To see what I mean, see `CalculateMemPoolAncestors`, which implements this sort of checking.

kallewoof commented:
@jimpo Thanks for spelling it out. To summarize:

1. The max ancestors and max descendants stuff is needlessly complex and can be simplified.
2. It is also broken.
3. It can be replaced with a single value (e.g. `max descendants`) that checks the descendant count of the top parent in the mempool.

Does that seem accurate?

kallewoof commented:
Delving into this a bit further, the ascendant vs descendant thing is quite different. There's usually only one ascendant per transaction except if several transactions were merged together, while there can be a tree of transactions as descendants.

I'm wondering if this should instead be "max unconfirmed group size", and default to the sum of the default values for max descendants/ascendants.

kallewoof commented:
Actually, I think the more straightforward fix here is to fix the description of max descendants, and to add it to the eligibility filter.

This PR can then be rewritten to take both max ancestors and max descendants. That would address your concerns, right?

kallewoof committed:
mempool: Add explicit max_descendants

TransactionWithinChainLimits would take a 'limit' and check it against ascendants and descendants. This is changed to take an explicit
max ancestors and max descendants value, and to test the corresponding value against its corresponding max.

kallewoof committed:
mempool: Fix max descendants check

The chain limits check for max descendants would check the descendants of the transaction itself even though the description for -limitdescendantcount says 'any ancestor'. This commit corrects the descendant count check by finding the top parent transaction in the mempool and comparing against that.

kallewoof commented:
@jimpo 

* 2d8748b adds a `max_descendants` to the eligibility class and uses it in the eligibility method in coin selection
* 4e742ba addresses the max descendants check to iterate over parents and using the top parent's descendant count. This is not exact, because a tx with multiple parents in the mempool will pick one of them and only use its values, but it's better than what we are doing right now.

Would you mind checking that out and seeing if it looks okay to you?

Empact commented on src/txmempool.cpp:
```diff
@@ -1055,11 +1055,25 @@ void CTxMemPool::TrimToSize(size_t sizelimit, std::vector<COutPoint>* pvNoSpends
+    ancestors = descendants = -1;
+    if (it != mapTx.end()) {
+        ancestors = it->GetCountWithAncestors();
+        descendants = (int64_t)CalculateDescendantMaximum(it);
```
Maybe add some bounds-checking asserts here ala `assert(int64_t(nCountWithDescendants) > 0);`

kallewoof replied:
I'm going to switch back to unsigned, so no need.

jimpo commented on src/wallet/wallet.cpp:
```diff
@@ -2467,8 +2467,11 @@ bool CWallet::OutputEligibleForSpending(const COutput& output, const CoinEligibi
-    if (!mempool.TransactionWithinChainLimit(output.tx->GetHash(), eligibility_filter.max_ancestors, eligibility_filter.max_descendants))
+    int64_t ancestors, descendants;
+    mempool.GetTransactionAncestry(output.tx->GetHash(), ancestors, descendants);
+    if (ancestors >= eligibility_filter.max_ancestors || descendants >= eligibility_filter.max_descendants) {
```
commit: Switch to GetTransactionAncestry() in OutputEligibleForSpending

I think these should be strict `>` checks, in which case you don't need to switch to signed results. This makes sense logically because ancestors and descendants counts includes the transaction queried, so if the count is at the maximum allowable, it is still valid to create one more in the chain.

kallewoof replied:
It means having to change the parameter for all the calls (bump by 1), but I think you're right. The -1 hack is bad. Fixing.

kallewoof replied:
I had to do one minor tweak, but otherwise the current values worked even with the change. I also switched from `nMaxChainLength` to using `max_ancestors` and `max_descendants` with their corresponding defaults. (See 6ab847bce76224ff46552dd647b542fa70350a91.)

jimpo commented on src/txmempool.cpp:
```diff
@@ -1055,11 +1055,22 @@ void CTxMemPool::TrimToSize(size_t sizelimit, std::vector<COutPoint>* pvNoSpends
+    for (;;) {
+        const setEntries& parents = GetMemPoolParents(top);
+        if (parents.size() == 0) break;
+        top = *parents.begin();
```
commit: mempool: Fix max descendants check

You actually need to recurse down all parents and take the max `GetCountWithDescendants` of all ancestors. A transaction can have multiple chains of ancestry with different descendants counts. This is also why it kind of makes sense to take the descendants limit as a parameter to this method rather than (or in addition to) returning it -- because it can limit the recursion.

kallewoof replied:
I was afraid that would be too resource intensive and potential for DoS. Doing a 'depth first optimistic pass' seemed like a good estimate.

jimpo replied:
Hmm, I understand the concern, but this doesn't match the advertised behavior of the flag. Also, the ancestor limit can be used to limit the depth of recursion. I'd like some more opinions on this.

kallewoof replied:
After thinking about it for a bit, I am not convinced it's particularly DoS vulnerable. I pushed a commit (to squash) that calculates it the way it is advertised.

jimpo reviewed (commented):
Definitely a big step in the right direction.

kallewoof committed:
wallet: Switch to using ancestor/descendant limits

Instead of combining the -limitancestorcount and -limitdescendantcount into a nMaxChainLength, this commit uses each one separately in the coin eligibility filters.

jimpo commented on src/wallet/wallet.cpp:
```diff
@@ -2592,7 +2592,7 @@ bool CWallet::SelectCoins(const std::vector<COutput>& vAvailableCoins, const CAm
         (m_spend_zero_conf_change && SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, CoinEligibilityFilter(0, 1, std::min((size_t)4, nMaxChainLength/3)), vCoins, setCoinsRet, nValueRet, coin_selection_params, bnb_used)) ||
         (m_spend_zero_conf_change && SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, CoinEligibilityFilter(0, 1, nMaxChainLength/2), vCoins, setCoinsRet, nValueRet, coin_selection_params, bnb_used)) ||
-        (m_spend_zero_conf_change && SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, CoinEligibilityFilter(0, 1, nMaxChainLength), vCoins, setCoinsRet, nValueRet, coin_selection_params, bnb_used)) ||
+        (m_spend_zero_conf_change && SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, CoinEligibilityFilter(0, 1, nMaxChainLength-1), vCoins, setCoinsRet, nValueRet, coin_selection_params, bnb_used)) ||
```
This underflows if nMaxChainLength is 0.

kallewoof replied:
Fixed (I set `nMaxChainLength` to `std::max(1, ...)`).

jimpo commented on src/wallet/wallet.cpp:
```diff
@@ -2583,16 +2583,17 @@ bool CWallet::SelectCoins(const std::vector<COutput>& vAvailableCoins, const CAm
-        (m_spend_zero_conf_change && SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, CoinEligibilityFilter(0, 1, nMaxChainLength-1), vCoins, setCoinsRet, nValueRet, coin_selection_params, bnb_used)) ||
+        (m_spend_zero_conf_change && SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, CoinEligibilityFilter(0, 1, std::min((size_t)4, max_ancestors/3), std::min((size_t)4, max_descendants/3)), vCoins, setCoinsRet, nValueRet, coin_selection_params, bnb_used)) ||
+        (m_spend_zero_conf_change && SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, CoinEligibilityFilter(0, 1, max_ancestors/2, max_descendants/2), vCoins, setCoinsRet, nValueRet, coin_selection_params, bnb_used)) ||
+        (m_spend_zero_conf_change && SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, CoinEligibilityFilter(0, 1, max_ancestors-1, max_descendants-1), vCoins, setCoinsRet, nValueRet, coin_selection_params, bnb_used)) ||
```
This underflows if max_ancestors or max_descendants is 0.

kallewoof replied:
Fixed (set both to `std::max(1, ...)`).

kallewoof committed:
wallet: Strictly greater than for ancestor caps

kallewoof committed:
mempool: Calculate descendant maximum thoroughly

jimpo commented:
utACK 6e56fef. Would be nice to have a unit test for `CTxMemPool::GetTransactionAncestry` though. :-)

kallewoof committed:
test: Add MempoolAncestryTests

kallewoof commented:
@jimpo I added tests in fcf1ed7 -- it would be great if you could check that the expected values matches what you would expect (checking the very last test is probably quickest, as each test simply builds on the previous one).

sdaftuar commented:
@kallewoof Sorry I missed you on IRC, but thanks for responding.

> <kallewoof> When doing coin selection on groups rather than individual outputs, the ancestor/descendant stuff becomes rather messy if you don't unify them to their corresponding maximums. In order to do that, you need the values, since they are tested against multiple limits.

I guess I don't understand how this relates because at first thought, I would have assumed that the only time we're spending unconfirmed outputs in the coin selection logic is if we are spending our own unconfirmed change, where address re-use (and therefore the grouping logic being introduced in #12257) should not be an issue at all.  Is there a case I'm overlooking?

kallewoof commented:
@sdaftuar Yes, I think most of the time the unspent trusted outputs will be your own and will probably not do address reuse, but I don't think it's guaranteed to always be that way.

Either way, once the coin selection is done per-group rather than per-output, it has to either iterate over all the outputs and check the limits each time, or it can cache it by storing the ancestry in the group struct itself. The latter seems like a more efficient solution and is enabled by this PR.

jimpo commented:
ACK e39efa8c9da85737393ca93bcda711774115702c

kallewoof commented:
Squashed fix-commits into 5ad55d2.

Empact commented on src/test/mempool_tests.cpp:
```diff
@@ -571,4 +571,182 @@ BOOST_AUTO_TEST_CASE(MempoolSizeLimitTest)
+        tx.vin[i].prevout.hash = inputs[i]->GetHash();
+        tx.vin[i].prevout.n = input_indices.size() > i ? input_indices[i] : 0;
+    }
+    for (size_t i = 0; i < output_values.size(); i++) {
```
nit: `++i`

Empact commented on src/txmempool.cpp:
```diff
@@ -1055,11 +1055,37 @@ void CTxMemPool::TrimToSize(size_t sizelimit, std::vector<COutPoint>* pvNoSpends
+        txiter candidate = candidates.back();
+        candidates.pop_back();
+        if (counted.find(candidate) != counted.end()) continue;
+        counted.insert(candidate);
```
nit: should be faster if you read the return value from the insertion, rather than find & insert.

kallewoof replied:
Didn't realize `set::insert` told you if the insertion happened. Nice!

Empact commented on src/txmempool.cpp:
```diff
@@ -1055,11 +1055,36 @@ void CTxMemPool::TrimToSize(size_t sizelimit, std::vector<COutPoint>* pvNoSpends
-bool CTxMemPool::TransactionWithinChainLimit(const uint256& txid, size_t chainLimit) const {
+uint64_t CTxMemPool::CalculateDescendantMaximum(txiter entry) const {
+    // find parent with highest descendant count
+    std::vector<txiter> candidates;
```
nit: more true to form to use a `std::stack` here, although I see they're not used otherwise in the codebase fwiw.

promag replied:
nit,
```cpp
std::vector<txiter> candidates = {entry};
```
and remove `push_back` below.

Empact commented:
utACK f77e1d3

promag commented on src/txmempool.cpp:
```diff
@@ -1055,11 +1055,36 @@ void CTxMemPool::TrimToSize(size_t sizelimit, std::vector<COutPoint>* pvNoSpends
 }
 
-bool CTxMemPool::TransactionWithinChainLimit(const uint256& txid, size_t chainLimit) const {
+uint64_t CTxMemPool::CalculateDescendantMaximum(txiter entry) const {
```
nit `const txiter& entry`.

promag reviewed (commented):
utACK f77e1d3.

sdaftuar commented:
Sorry I haven't dug into this change myself, but might this have a material performance impact on wallets with many unconfirmed coins, if the mempool is full?

kallewoof commented:
@sdaftuar I honestly doubt it, unless they have a TON of ancestors in the mempool. I was initially hesitant to make the ancestor check "thorough" but couldn't come up with a reason why it would ever be the case that you have enough transactions in the mempool to make the ancestor check bog down your system significantly.
